src/main/java/org/apache/sling/xss/impl/xml/Policy.java

```diff
@@ -0,0 +1,391 @@
+                String attrName = name.toLowerCase().substring(0, name.length() - 1);
+                dynamicAttributes.put(attrName, toAdd);
+            } else
+                throw new PolicyException("Dynamic attribute '" + name
+                        + "' was not defined in <common-attributes>");
+        }
+    }
+
+    private List<Regexp> getAllowedRegexps(List<Regexp> nameAndRegexpsList) {
+        List<Regexp> allowedRegExp = new ArrayList<>();
+        if (nameAndRegexpsList != null && nameAndRegexpsList.size() > 0) {
+            for (Regexp regExpNode : nameAndRegexpsList) {
+                String regExpName = regExpNode.getName();
+                String value = regExpNode.getValue();
+
+                if (regExpName != null && regExpName.length() > 0) {
+                    allowedRegExp
+                            .add(new Regexp(regExpNode.getName(),
+                                    commonRegularExpressions.get(regExpName).toString()));
+                } else if (value != null) {
+                    allowedRegExp.add(new Regexp(regExpName, value));
+                }
+            }
+        }
+        return allowedRegExp;
+    }
+
+    private void parseTagRules(List<Tag> root) throws PolicyException {
+        if (root == null)
+            return;
+
+        for (Tag tagNode : root) {
+            String name = tagNode.getName();
+            String action = tagNode.getAction();
+
+            List<Attribute> attributeList = tagNode.getAttributeList();
+            List<Attribute> tagAttributes = getTagAttributes(attributeList, name);
+            Tag tag = new Tag(name, action, tagAttributes);
+
+            tagRules.put(name.toLowerCase(), tag);
+        }
+    }
+
+    private List<Attribute> getTagAttributes(List<Attribute> attributeList, String tagName)
+            throws PolicyException {
+
+        List<Attribute> tagAttributes = new ArrayList<>();
+        if (attributeList != null && attributeList.size() > 0) {
+            for (Attribute attribute : attributeList) {
+                Attribute newAttribute;
+                String attributeName = attribute.getName().toLowerCase();
+                List<Regexp> regexps = attribute.getRegexpList();
+                List<Literal> literals = attribute.getLiteralList();
+                String onInvalid = attribute.getOnInvalid();
+                String description = attribute.getDescription();
+
+                // attribute has no children
+                if (regexps == null && literals == null) {
+
+                    Attribute commonAttribute = commonAttributes.get(attributeName);
+                    if (commonAttribute != null) {
+                        // creates a new Attribut with the fetched Attribute informations if not
+                        // available
+                        newAttribute = new Attribute(attributeName,
+                                regexps != null && regexps.size() != 0 ? regexps : commonAttribute.getRegexpList(),
```

Comment:
Please use empty collections as defaults to avoid these null checks.

Reply:
Resolved: 91f4846ed54726f6d095d45d819f3fdc7a480199